Store TLS version in SSL session structure

hanno-arm opened the pull request:
Instances of `mbedtls_ssl_session` represent data enabling session resumption.

With the introduction of TLS 1.3, the format of this data changes. We therefore
need TLS-version field as part of `mbedtlsl_ssl_session` which allows distinguish
1.2 and 1.3 sessions.

This commit introduces such a TLS-version field to mbedtls_ssl_session.

The change has a few ramifications:

- Session serialization/deserialization routines need to be adjusted.

  This is achieved by adding the TLS-version after the header of
  Mbed TLS version+config, and by having the subsequent structure
  of the serialized data depend on the value of this field.

  The details are described in terms of the RFC 8446 presentation language.

  The 1.2 session (de)serialization are moved into static helper functions,
  while the top-level session (de)serialization only parses the Mbed TLS
  version+config header and the TLS-version field, and dispatches according
  to the found version.

  This way, it will be easy to add support for TLS 1.3 sessions in the future.

- Tests for session serialization need to be adjusted

- Once we add support for TLS 1.3, with runtime negotiation of 1.2 vs. 1.3,
  we will need to have some logic comparing the TLS version of the proposed session
  to the negotiated TLS version. For now, however, we only support TLS 1.2,
  and no such logic is needed. Instead, we just store the TLS version in the
  session structure at the same point when we populate mbedtls_ssl_context.minor_ver.

The change introduces some overlap between `mbedtls_ssl_session.minor_ver` and
`mbedtls_ssl_context.minor_ver`, which should be studied and potentially resolved.
However, with both fields being private and explicitly marked so, this can happen
in a later change.

Fixes #4802

gilles-peskine-arm reviewed (changes requested):
Looks mostly good, except I don't understand why the context and session structures now both have a minor version field. It's no big deal right now since the library only supports a single version, but if we go down that route adding a version field is pointless. When the library starts supporting multiple version, we'll have to do extra work keeping those fields synchronized.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4550,44 +4550,62 @@ static unsigned char ssl_serialized_session_header[] = {
+ *  struct {
+ *
+ *    opaque mbedtls_version[3];   // major, minor, patch
+ *    opaque session_format[2];    // version-specific 16-bit field determining
```
1. Good point, is shoud library-version-specific.
2. It is mostly for the code-structure: `mbedtls_version` and `session_format` are determined at compile-time, so we can parse and validate it with a simple `memcpy`. Protocol-specific fields follow. I considered integrating the protocol-version here, but found it preferable to keep the structure as we have it now.

hanno-arm replied:
Fixed 1.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -927,6 +927,8 @@ struct mbedtls_ssl_session
 
     unsigned char exported;
 
+    unsigned char MBEDTLS_PRIVATE(minor_ver);    /*!< The TLS version used in the session. */
```
To keep the change small and allow us to move forwards - that's the only reason. I think a good time to revisit this is #4823.

gilles-peskine-arm replied:
Ok, but please:
* Add a comment to the documentation of both fields mentioning the temporary duplication.
* Record the duplication as technical debt to resolve, either in #4823 or a dedicated issue.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4550,44 +4550,62 @@ static unsigned char ssl_serialized_session_header[] = {
+ *    opaque ticket<0..2^24-1>;       // length 0 means no ticket
+ *    uint32 ticket_lifetime;
+ *    uint8 mfl_code;                 // up to 255 according to standard
+ *    uint8 trunc_hmac;               // 0 or 1
```
Sure!

hanno-arm replied:
Fixed.

hanno-arm commented:
Thank you @gilles-peskine-arm for your review! I have addressed your comments, and tried to explain why I prefer not to try to consolidate `mbedtls_ssl_context.minor_ver` and `mbedtls_ssl_session.minor_ver` at this point.

mpg commented on library/ssl_tls.c:
```diff
@@ -4560,7 +4560,7 @@ static unsigned char ssl_serialized_session_header[] = {
  *                                 // the structure of mbedtls_ssl_session.
  *
- *    uint8_t minor_ver;           // Possible values:
+ *    uint8_t minor_ver;           // Protol-version. Possible values:
```
Typo: protocol

hanno-arm replied:
Fixed!

mpg reviewed (changes requested):
Looks good to me except for a typo in a comment.

mpg reviewed (approved):
LGTM
